card/mod/standard/set/all/links.rb

```diff
@@ -80,79 +71,126 @@ def view_link text, view, opts={}
     opts[:path_opts] ||= { view: :related }
     opts[:path_opts][:related] = { name: "+#{name}" }
     opts[:path_opts][:related].merge! opts[:related_opts] if opts[:related_opts]
     view_link(opts[:text] || name, :related, opts)
   end
 
-  def path opts={}
-    if opts[:action] == :new && opts[:type] &&
-       !(opts[:name] || opts[:card] || opts[:id])
-      opts.delete(:action)
-      base = "new/#{opts.delete(:type)}"
-    else
-      name = opts.delete(:name) || card.name
-      base = opts[:action] ? "card/#{opts.delete :action}/" : ""
+  # @param opts [Hash]
+  # @option opts [Symbol] :action card action (:create, :update, :delete)
+  # @option opts [Integer, String] :id
+  # @option opts [String, Card::Name] :name
+  # @option opts [String] :type
+  # @option opts [Hash] :card
+  # @param mark_type [Symbol] defaults to :id
+  def path opts={}, mark_type=:id
+    base = new_cardtype_path(opts) || standard_path(opts, mark_type)
+    query = path_query(opts)
+    internal_url base + query
+  end
 
-      opts[:no_id] = true if [:new, :create].member? opts[:action]
-      # generalize. dislike hardcoding views/actions here
+  def new_cardtype_path opts
+    return unless opts[:action] == :new
+    opts.delete :action
+    return unless opts[:type] && !opts[:name] && !opts[:card] && !opts[:id]
+    "new/#{opts.delete :type}"
+  end
 
-      linkname = name.to_name.url_key
-      unless name.empty? || opts.delete(:no_id)
-        base += (opts[:id] ? "~#{opts.delete :id}" : linkname)
-      end
+  def standard_path opts, mark_type
+    standardize_action! opts
+    path_action(opts[:action]) + path_mark(opts, mark_type)
+  end
 
-      process_path_card_opts opts, name, linkname
+  def path_action action
+    case action
+    when :create then "card/#{action}/"
+    # sometimes create action has no mark,
+    # but /create refers to a card named "create"
+    when nil     then ""
+    else              "#{action}/"
     end
+  end
 
-    query = opts.empty? ? "" : "?#{opts.to_param}"
-    internal_url(base + query)
+  def standardize_action! opts
+    return if [:create, :update, :delete].member? opts[:action]
+    opts.delete :action
   end
 
-  def internal_url relative_path
-    card_path relative_path
+  def path_mark opts, mark_type
+    case mark_type
+    when :id       && (id = path_id opts)        then "~#{id}"
+    when :codename && (codename = card.codename) then ":#{codename}"
+    else (opts[:name] || card.name).to_name.url_key
+    end
   end
 
-  def interpret_href href
-    href.is_a?(Hash) ? path(href) : href
+  def path_id opts
+    id = opts.delete :id
+    id if id.present?
   end
 
-  def process_path_card_opts opts, name, linkname
-    opts[:card] ||= {}
-    if opts.delete(:known) == false && name.present? && name.to_s != linkname
-      opts[:card][:name] = name
+  def path_query opts
+    finalize_card_opts opts.delete(:card), opts
+    opts.delete :action
+    opts.empty? ? "" : "?#{opts.to_param}"
```

Comment:
no, not yet, but I don't really like the idea of `opts: { params: {} }`

I think `opts` should basically be `params` and anything that doesn't fit that should be separate.  In fact, I might ultimately get it to `def path params, opts`.

path should map very cleanly to routes.  :action, :card, and :id are pretty normal params but with weird roles in routes.  I want to switch :id to :mark in both routes.rb and #path, and it may help solve the name problem if I can get :name merged into :mark. 

:type is the other weird one.  will discuss above in your other comment.